File: src/main/java/frc/robot/DistributionHub/PowerDistributionHub.java
package frc.robot.DistributionHub;

import edu.wpi.first.wpilibj.PowerDistribution;
import edu.wpi.first.wpilibj.PowerDistribution.ModuleType;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;

public class PowerDistributionHub {

    public int SelectedChannelNumber = 0; // number of slot on PDH for monitoring current through individaul channel
    
    PowerDistribution MainPowerDistributionHub = new PowerDistribution(1, ModuleType.kRev); // CAN ID must be 1 for REV PDH. If using CTRE PDP, use 0.

    public void getVoltage() {// get the voltage coming from the battery of the robot to the PDH
        double Voltage = MainPowerDistributionHub.getVoltage();
        SmartDashboard.putNumber("System Voltage", Voltage);
    }

    public void getTemperatureInC() { // get temperature of power distribution hub in celsius
        double TemperatureInC = MainPowerDistributionHub.getTemperature();
        SmartDashboard.putNumber("PDH Temp", TemperatureInC);
    }

    public void getTotalCurrent() { // get current coming from battery
        double TotalCurrent = MainPowerDistributionHub.getTotalCurrent();
        SmartDashboard.putNumber("System Current", TotalCurrent);
    }

    public void getIndividualCurrent() { // get current of each individual channel
        double selectedPortCurrent = MainPowerDistributionHub.getCurrent(SelectedChannelNumber); 
    }

    public void LEDs() { 
        MainPowerDistributionHub.setSwitchableChannel(true);
    }
}
    
    

File: subsystems/power_distribution_hub.py
"""
Power Distribution Hub monitoring subsystem.

Publishes voltage, temperature, total current, and per-channel current to
SmartDashboard.
"""

import wpilib
from commands2 import Subsystem
from wpilib import SmartDashboard

from constants import PDH_ID


class PowerDistributionHub(Subsystem):
    def __init__(self):
        super().__init__()
        self.pdh = wpilib.PowerDistribution(PDH_ID, wpilib.PowerDistribution.ModuleType.kRev)

    def publish_telemetry(self):
        """Publish PDH data to SmartDashboard."""
        SmartDashboard.putNumber("PDH/Voltage", self.pdh.getVoltage())
        SmartDashboard.putNumber("PDH/Temperature", self.pdh.getTemperature())
        SmartDashboard.putNumber("PDH/TotalCurrent", self.pdh.getTotalCurrent())
        for ch in range(24):
            SmartDashboard.putNumber(f"PDH/CH{ch} Current", self.pdh.getCurrent(ch))
